### PR TITLE
Visual geometry key

### PR DESCRIPTION
**Pull request by RobBuchananCompPhys**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce(Bug fix, feature, docs update, ...)?** 
This is a feature and accompanied docs update

* **What is the current behaviour (You can also link to an open issue here)?**

Currently, the IDF JSON can be used to specify a mesh file for an instrument component. The designer can be used to edit this property, which is required. 

* **What is the new behaviour (if this is a feature change)?**
With this feature we have the option to supply a geometry object as a sub-property of the visual component instead of the mesh property. 

The geometry can include either a primitive (box, plane or sphere) or the path to a mesh file.

A geometry is required if there is no mesh for the visual, and vice-versa (although the geometry property will take precedence for new IDF files, and the mesh field will remain supported for backward compatibility). The instrument schema has therefore been updated to respond with a 'geometry required' warning if the mesh and geometry properties are absent, but will accept a mesh field with no geometry. 

The designer tool in the SScanss-2 editor can integrate with instrument JSON and allows values to be specified for the possible geometry types.

* **Does this PR introduce a breaking change (What changes might users need to make in their application due to this PR)?**
None - the visual mesh property is still supported.

### CODE REVIEW

File: sscanss/editor/designer.py
Comment (StephenNneji) [inline sscanss/editor/designer.py:1]: Screenshot 2023-09-13 160827
please adjust the combobox to stretch like the others

Screenshot 2023-09-13 160912
Please adjust the dimensions/radius column to line up with Type

**Review by StephenNneji — COMMENTED** [review]
Good Job, please address comments

File: sscanss/core/instrument/create.py [hunk 1]
```diff
@@ -217,19 +221,49 @@ def read_visuals(visuals_data, path=''):
...
+                return
+
+            geom_type = str(geometry_data['type']).capitalize()
+
+            if geom_type == VisualGeometry.Mesh.value:
+                mesh = read_3d_model(pathlib.Path(path).joinpath(visuals_data['geometry'].get('path', '')).as_posix())
+
+            elif geom_type == VisualGeometry.Box.value:
+                size = geometry_data.get('size', [])
+                if len(size) != 3:
```
[the remark this answers is not in the document]
Comment (RobBuchananCompPhys) [reply sscanss/core/instrument/create.py:245]: Should the exception be raised in the editor designer code, or here in `create_visuals`?

File: tests/test_editor.py
[the remark this answers is not in the document]
Comment (RobBuchananCompPhys) [reply tests/test_editor.py:1]: Sounds good

File: sscanss/editor/designer.py [hunk 2]
```diff
@@ -235,13 +248,280 @@ def value(self):
...
+                    menu_item.layout().deleteLater()
+        self.setFilePicker()
+
+    def setMenu(self, type):
+        """Sets up the menu input line edits depending on the currently selected geometry
+        
+        :param type: string representation of the currently selected geometry type
+        :type type: Literal['Box', 'Sphere', 'Plane', 'Mesh']
+        :return: returns none if geometry not recognized
+        :rtype: none
```
[the remark this answers is not in the document]
Comment (RobBuchananCompPhys) [reply sscanss/editor/designer.py:321]: I agree, makes sense to it that way

File: sscanss/core/instrument/create.py [hunk 3]
```diff
@@ -208,7 +211,8 @@ def read_detector_description(instrument_data, positioners, path=''):
 
 
 def read_visuals(visuals_data, path=''):
-    """Creates Mesh object from a visuals description
```
Comment (StephenNneji) [inline sscanss/core/instrument/create.py:211]: A lot of the validation is already done by the schema and will never get called, please see my suggested refactor below
```def read_visuals(visuals_data, path=''):
    """Creates Mesh object from a visuals description using a file if a mesh path supplied, 
    else the specified primitive geometry is used

    :param visuals_data: visuals description
    :type visuals_data: Optional[Dict]
    :param path: directory of the instrument description file
    :type path: str
    :return: mesh
    :rtype: Optional[Mesh]
    """
    if visuals_data:
        pose = visuals_data.get('pose', DEFAULT_POSE)
        pose = matrix_from_pose(pose)
        mesh_colour = visuals_data.get('colour', DEFAULT_COLOUR)

        if visuals_data.get('mesh'):
            mesh = read_3d_model(pathlib.Path(path).joinpath(visuals_data.get('mesh', '')).as_posix())

        geometry_data = visuals_data['geometry']
        geom_type = geometry_data['type'].capitalize()

        if geom_type == VisualGeometry.Mesh.value:
            mesh_file = geometry_data['path']
            mesh = read_3d_model(pathlib.Path(path).joinpath(mesh_file).as_posix())

        elif geom_type == VisualGeometry.Box.value:
            x, y, z = geometry_data['size']
            mesh = create_cuboid(x, z, y)

        elif geom_type == VisualGeometry.Plane.value:
            x, y = geometry_data['size']
            mesh = create_plane(Plane.fromCoefficient(1, 1, 0, 0), x, y)

        elif geom_type == VisualGeometry.Sphere.value:
            radius = geometry_data['radius']
            mesh = create_sphere(radius)

        mesh.transform(pose)
        mesh.colour = Colour(*mesh_colour)
        return mesh
```

File: sscanss/editor/designer.py [hunk 4]
```diff
@@ -235,13 +233,285 @@ def value(self):
...
+            self.y_input = create_validated_line_edit(3, str(y))
+            self.y_input.textChanged.connect(lambda y: form.update({"y": safe_get_value([y], 0, self.default_size)}))
+            self.menu.addWidget(QtWidgets.QLabel('Size: '), 1, 0)
+            self.menu.addLayout(xy_hbox_layout(self.x_input, self.y_input), 1, 1)
+
+        elif type == VisualGeometry.Mesh.value:
+            self.menu.addWidget(QtWidgets.QLabel('Mesh: '), 3, 0)
+            self.menu.addWidget(self.file_picker, 3, 1)
+
+        else:
```
Comment (StephenNneji) [inline sscanss/editor/designer.py:359]: This function still returns None for the other branches it should return true

**Review by StephenNneji — COMMENTED** [review]
Please address comments

**Review by StephenNneji — APPROVED** [review]

**Pull request merged**